velox/core/PlanNode.cpp

```diff
@@ -345,6 +345,68 @@ PlanNodePtr AggregationNode::create(const folly::dynamic& obj, void* context) {
       deserializeSingleSource(obj, context));
 }
 
+namespace {
+RowTypePtr getSparkExpandOutputType(
+    const std::vector<std::vector<TypedExprPtr>>& projectSets,
+    const std::vector<std::string>& names) {
+  std::vector<std::string> outputs;
+  outputs.reserve(names.size());
+  std::vector<TypePtr> types;
+  types.reserve(names.size());
+  for (int32_t i = 0; i < names.size(); ++i) {
+    outputs.push_back(names[i]);
+    auto expr = projectSets[0][i];
+    types.push_back(expr->type());
+  }
+
+  return ROW(std::move(outputs), std::move(types));
+}
+} // namespace
+
+ExpandNode::ExpandNode(
+    PlanNodeId id,
+    std::vector<std::vector<TypedExprPtr>> projectSets,
+    std::vector<std::string> names,
+    PlanNodePtr source)
+    : PlanNode(std::move(id)),
+      sources_{source},
+      outputType_(getSparkExpandOutputType(projectSets, names)),
+      projectSets_(std::move(projectSets)),
+      names_(std::move(names)) {}
```

Comment:
@mbasmanova  For item four, there is no guarantee that projectSets[i][n]->type() == projectSets[j][n]->type() because if projectSets[i][n] is null and projectSets[j][n] is not null, their types will be different.

Reply:
@JkSelf Would you clarify what are possible types when one of the projections is null?

Reply:
@mbasmanova I had a misunderstanding regarding item four earlier. Have added the type check.